apps/src/dropletUtils.js

```diff
@@ -157,41 +159,43 @@ standardConfig.categories = {
 };
 
 /**
- * @param codeFunctions
+ * @param codeFunctions {object} A collection of named key/value pairs
+ *   key is a block name from dropletBlocks or standardBlocks
+ *   value is an object that can be used to override block defaults
  * @param {DropletConfig} dropletConfig
  * @param {DropletConfig} otherConfig optionally used to supply a standardConfig
  *  object which is not app specific. It will be used first, then overriden
  *  by the primary dropletConfig if there is overlap between the two.
  * @param paletteOnly boolean: ignore blocks not in codeFunctions palette
- * @returns {Array}
+ * @returns {Array<DropletBlock>}
  */
-function mergeFunctionsWithConfig(codeFunctions, dropletConfig, otherConfig, paletteOnly) {
+function filteredBlocksFromConfig(codeFunctions, dropletConfig, otherConfig, paletteOnly) {
   if (!codeFunctions || !dropletConfig || !dropletConfig.blocks) {
     return [];
   }
 
-  var merged = [];
-
-  var blockSets = [ dropletConfig.blocks ];
+  var blocks = [];
   if (otherConfig) {
-    blockSets.splice(0, 0, otherConfig.blocks);
+    blocks = blocks.concat(otherConfig.blocks);
   }
-
-  // codeFunctions is an object with named key/value pairs
-  //  key is a block name from dropletBlocks or standardBlocks
-  //  value is an object that can be used to override block defaults
-  for (var s = 0; s < blockSets.length; s++) {
-    var set = blockSets[s];
-    for (var i = 0; i < set.length; i++) {
-      var block = set[i];
-      if (!paletteOnly || block.func in codeFunctions) {
-        // We found this particular block, now override the defaults with extend
-        merged.push($.extend({}, block, codeFunctions[block.func]));
-      }
+  blocks = blocks.concat(dropletConfig.blocks);
+
+  var docFunctions = {};
+  blocks.forEach(function (block) {
+    // For cases where we use a different block for our tooltips, make sure that
+    // the target block ends up in the list of blocks we want
+    var docFunc = block.docFunc;
+    if (docFunc && !(docFunc in codeFunctions)) {
+      docFunctions[docFunc] = null;
     }
-  }
-
-  return merged;
+  });
+
+  return blocks.filter(function (block) {
+    return !paletteOnly || block.func in codeFunctions || block.func in docFunctions;
+  }).map(function (block) {
+    // We found this particular block, now override the defaults with extend
+    return $.extend({}, block, codeFunctions[block.func]);
+  });
 }
```

Comment:
This method looks like a lot of delta, but most of it is doing the same thing as before, but written differently. The key portion is the generation of the docFunctions hash, which we use to include referenced blocks.